src/v0/sources/shopify/util.js

```diff
@@ -111,11 +119,204 @@ const setAnonymousId = (message) => {
+          message.setProperty('userId', 'shopify-admin');
+        }
+        return;
+      }
+      cartToken = message.properties?.cart_token;
+      break;
+    /*
+     * we dont have cart_token for carts_create and update events but have id and token field
+     * which later on for orders become cart_token so we are fethcing cartToken from id
+     */
+    case SHOPIFY_TRACK_MAP.carts_create:
+    case SHOPIFY_TRACK_MAP.carts_update:
+      cartToken = message.properties?.id;
+      break;
+    // https://help.shopify.com/en/manual/orders/edit-orders -> order can be edited through shopify-admin only
+    // https://help.shopify.com/en/manual/orders/fulfillment/setting-up-fulfillment -> fullfillments wont include cartToken neither in manual or automatiic
+    case SHOPIFY_TRACK_MAP.orders_edited:
+    case SHOPIFY_TRACK_MAP.fulfillments_create:
+    case SHOPIFY_TRACK_MAP.fulfillments_update:
+      if (!message.userId) {
+        message.setProperty('userId', 'shopify-admin');
+      }
+      return;
+    default:
+  }
+  let anonymousIDfromDB = await anonymousIdCache.get(cartToken, () => null); // check if anonymousId is present in cachewith cartToken
+  if (anonymousIDfromDB === null) {
+    const executeStartTime = Date.now();
+    const redisVal = await RedisDB.getVal(`${cartToken}`);
+    stats.timing('redis_get_latency', executeStartTime, {
+      ...metricMetadata,
+    });
+    stats.increment('shopify_redis_get_data', {
+      ...metricMetadata,
+      timestamp: Date.now(),
+    });
+    if (redisVal !== null) {
+      anonymousIDfromDB = redisVal.anonymousId;
+    }
+    if (!isDefinedAndNotNull(anonymousIDfromDB)) {
+      /* this is for backward compatability when we don't have the redis mapping for older events
+      we will get anonymousIDFromDb as null so we will set UUID using the session Key */
+      anonymousIDfromDB = cartToken;
+    }
+  }
+  message.setProperty('anonymousId', anonymousIDfromDB);
+};
+
+/**
+ * It compares the line items element wise and checks the properties parameter for every element
+ * as for duplicate ones it may be same.
+ * It returns true if every item matches.
+ * Example:
+ * Input :
+ * prevLineItems: [
+ * {...,
+ * properties:null,
+ * ...
+ * },
+ * {...,
+ * properties:null,
+ * ...
+ * }]
+ *
+ * newLineItems: [
+ * {...,
+ * properties:{},
+ * ...
+ * },
+ * {...,
+ * properties:{},
+ * ...
+ * }]
+ *
+ * This above example can shwo you the need of this function
+ * that is to check for properties parameter as
+ * in some cases it is {} and in others as null
+ * which makes it difficult to comapre the json
+ * @param {*} prevLineItems
+ * @param {*} newLineItems
+ * @returns
+ */
+// https://shopify.dev/docs/api/liquid/objects/line_item#line_item-properties
+const isContainingSameLineItems = (prevLineItems, newLineItems) => {
+  let noOfItems = prevLineItems.length;
+  while (noOfItems > 0) {
+    const modifiedPrev = _.cloneDeep(prevLineItems[noOfItems - 1]);
```

Comment:
JSON stringify used, need clone Deep?

Reply:
Yess as you can see we are manipulating the cloned variables before stringifying it.